Fix another edge case with splitting code into lines

allejo opened the pull request:
The `.hljs-tag` here spans two lines, however, the `text()` XPath function will trim the newline so the XPath that was introduced in #79 doesn't work.

**Original Code**

```html
<?xml version="1.0" encoding="utf-8" ?>
               <tag a="t"
                        b="t">
                 </tag>
```

**Highlighted Result**

```html
<span class="hljs-meta">&lt;?xml version="1.0" encoding="utf-8" ?&gt;</span>
               <span class="hljs-tag">&lt;<span class="hljs-name">tag</span> <span class="hljs-attr">a</span>=<span class="hljs-string">"t"</span>
                        <span class="hljs-attr">b</span>=<span class="hljs-string">"t"</span>&gt;</span>
                 <span class="hljs-tag">&lt;/<span class="hljs-name">tag</span>&gt;</span>
```

~~I have tried changing the XPath to use `.` instead of `text()`, however, that will match deeply nested newlines like those seen in #79 and multiply the number of "results" this XPath returns relative to the depth of the span that contains it.~~

~~Running the 9.18.1.5 release splits the above snippet in \~29 seconds for 200000 iterations using `text()` (however, it's incorrect in its splitting); replacing `text()` with `.` results in a time of \~34 seconds in the exact scenario but splits correctly.~~

~~This PR without XPath and manually searching for newlines inside of spans comes in at \~34 seconds for 200000 iterations as well.~~

---

I have taken a step back and rewritten this function entirely to no longer parse the HTML into a DOM object. I am now using pure regular expressions to do our best in figuring out which tags span multiple lines.

The solution in this PR functions as such:

1. With the highlighted code snippet result above, I split things on `\R` and I count how many opening and closing span tags there are on each line.
2. If we find an even amount of closing and opening tags, we move on to the next line
3. If there are an uneven amount of closing tags with opening tags, that means we found a span tag that spans multiple lines.
4. I remove all complete `<span>...</span>` tags from a copy of the current line
5. I then match all of the opening span tags and save them for our next line
6. I then count how many open tags there are and put them at the end of the current line
7. In the next line, if I have anything saved from step 5, then I add them as a prefix to the current line in this iteration

Without parsing the HTML into a DOM object, I've decreased the speed of 200000 iterations down to ~10 seconds.

---

See https://github.com/westonruter/syntax-highlighting-code-block/issues/246

/cc @westonruter

westonruter commented on HighlightUtilities/functions.php:
```diff
@@ -222,11 +222,25 @@ function splitCodeIntoArray($html)
 
     /** @var \DOMElement $span */
     foreach ($spans as $span) {
+        if ($span->hasChildNodes()) {
+            $hasNewlines = false;
+
+            foreach ($span->childNodes as $child) {
+                if ($child->nodeType === XML_TEXT_NODE && preg_match('/\R/', $child->textContent)) {
+                    $hasNewlines = true;
+                    break;
+                }
+            }
+
+            if (!$hasNewlines) {
+                continue;
+            }
+        }
```
Wouldn't this be the same as:

```suggestion
        if ($span->hasChildNodes() && ! preg_match('/\R/', $span->textContent)) {
            continue;
        }
```

Since `DOMElement::$textContent` includes the concatenation of all descendant text nodes.

allejo replied:
In the original attempt, this was not quite the case, by using `DOMElement::$textContent` it would include child `<span>`s. I wanted to find child nodes that were purely text and not additional spans.

But I wrote a different implementation dropping the DOM parser entirely.

allejo commented:
@westonruter I've rewritten this function entirely (again) and added two unit tests for splitting things correctly based on the reported issue. Could you take a look and/or if you have any more snippets to share?

Here's what WP looks like with this PR

![image](https://user-images.githubusercontent.com/1246453/102752299-327c0300-431e-11eb-9630-3a960d7a06e0.png)

westonruter commented:
> Without parsing the HTML into a DOM object, I've decreased the speed of 200000 iterations down to ~10 seconds.

And in `master` it is currently 29 seconds?

allejo commented:
> > Without parsing the HTML into a DOM object, I've decreased the speed of 200000 iterations down to ~10 seconds.
> 
> 
> 
> And in `master` it is currently 29 seconds?

Correct

westonruter commented:
So ~3X faster. Nice!

westonruter commented on HighlightUtilities/functions.php:
```diff
@@ -196,56 +196,59 @@ function getLanguageDefinitionPath($name)
-        $openingTags = '';
-        $curr = $span;
+        // Find how many opening `<span>` tags exist on this line
+        preg_match_all('/<span[^<>]+>/u', $rawLine, $regexWorkspace);
```
In order to match `<span>`:

```suggestion
        preg_match_all('/<span[^>]*+>/u', $rawLine, $regexWorkspace);
```

This also uses a “possessive quantifier” to prevent backtracking which is unnecessary, I think.

And I think `[^>]` can be used instead of `[^<>]` to match the pattern below?

allejo replied:
Good catch, I for some reason thought `<` could end up in the attributes some how but that would need to be escaped anyways

westonruter replied:
Actually, you're right that `<` can be in attributes, but so also can `>`. See https://jsbin.com/vitexad/edit?html,output

> ![image](https://user-images.githubusercontent.com/134745/102855594-c8289880-43d9-11eb-82cb-96a28fda201c.png)

This _works_ although and it is even valid HTML: 

> ![image](https://user-images.githubusercontent.com/134745/102855523-a6c7ac80-43d9-11eb-8bb0-27a8129893e5.png)

I think the question is whether or not `<` or `>` will ever show up in the `span` tags that highlight.php generates.

allejo replied:
I don't _think_ `<` and `>` would ever appear as class names since that's the only attribute highlight.php should ever generate; unless there's some really weird masochistic language grammar somewhere. I think it should be safe to ignore in this case

westonruter replied:
Yeah, that makes sense.

westonruter commented on HighlightUtilities/functions.php:
```diff
@@ -196,56 +196,59 @@ function getLanguageDefinitionPath($name)
-        $html = str_replace($renderedSpan, $finished, $html);
+        // Find all of the complete `<span>` tags and remove them from a working
+        // copy of the line. Then we'll be left with just opening tags.
+        $workingLine = preg_replace('/<span[^<]+>[^<>]+<\/span>/u', '', $rawLine);
```
Shouldn't this be able to match an empty `span`?

```suggestion
        $workingLine = preg_replace('/<span[^<]*+>[^<]*+<\/span>/u', '', $rawLine);
```

allejo replied:
I edited your suggestion to invert the `[^<]` to `[^>]` so I could batch commit via GH; was an original typo on my part

westonruter commented on HighlightUtilities/functions.php:
```diff
@@ -196,56 +196,59 @@ function getLanguageDefinitionPath($name)
+        // Find all of the complete `<span>` tags and remove them from a working
+        // copy of the line. Then we'll be left with just opening tags.
+        $workingLine = preg_replace('/<span[^<]+>[^<>]+<\/span>/u', '', $rawLine);
+        preg_match_all('/<span[^>]+>/u', $workingLine, $regexWorkspace);
```
```suggestion
        preg_match_all('/<span[^>]*+>/u', $workingLine, $regexWorkspace);
```
